File: ios/Sources/Web2AppSDK/AttributionResolver.swift
import Foundation

/// Резолв `guid` из carrier-token (deep_link_value / Install Referrer) или email.
///
/// ⚠ **POC-1 boundary:** SDK НЕ парсит внутренности MMP-SDK — интегратор передаёт готовый
/// `deep_link_value` из СВОЕГО AppsFlyer/Adjust callback в `Web2App.identify(deepLinkValue:)`.
/// POC-1 подтверждает, что MMP реально ДОСТАВЛЯЕТ `deep_link_value` на реальном iOS-девайсе
/// (adjust/ios_sdk#752, iOS17/18). До POC iOS-ветка ship-blocked — но код резолва токена
/// (ниже) POC-независим: как только token на руках, резолв в guid стабилен.
///
/// Интеграция MMP (talking-point для integration-doc):
/// ```swift
/// // AppsFlyer:
/// func onConversionDataSuccess(_ data: [AnyHashable: Any]) {
///     guard (data["af_status"] as? String) == "Non-organic",
///           (data["is_first_launch"] as? Bool) == true,
///           let token = data["deep_link_value"] as? String else {
///         Web2App.identify(deepLinkValue: nil) { ... } // → email-fallback
///         return
///     }
///     Web2App.identify(deepLinkValue: token) { ... }
/// }
/// ```
struct AttributionResolver {
    let config: Web2AppConfig

    /// carrier-token → guid: `GET /public/handoff/resolve?code=<token>` (стабильный, WEB-433).
    func resolveToken(_ token: String, completion: @escaping (Result<String, Web2AppError>) -> Void) {
        var comps = URLComponents(
            url: config.baseUrl.appendingPathComponent("public/handoff/resolve"),
            resolvingAgainstBaseURL: false
        )
        comps?.queryItems = [URLQueryItem(name: "code", value: token)]
        guard let url = comps?.url else { return completion(.failure(.resolveFailed)) }
        Self.fetchGuid(url: url, method: "GET", body: nil, completion: completion)
    }

    /// email-recovery запрос (шаг 1): `POST /public/handoff/email-recovery/request`
    /// {projectId, email} → 204. Сервер шлёт magic-link; guid придёт на шаге 2, когда юзер
    /// откроет ссылку (code) → resolveToken. Контракт подтверждён: public-handoff.controller.
    func requestEmailRecovery(
        _ email: String,
        completion: @escaping (Result<Void, Web2AppError>) -> Void
    ) {
        let url = config.baseUrl.appendingPathComponent("public/handoff/email-recovery/request")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try? JSONSerialization.data(withJSONObject: [
            "projectId": config.projectId,
            "email": email,
        ])
        URLSession.shared.dataTask(with: req) { _, resp, err in
            if let err { return completion(.failure(.network(err.localizedDescription))) }
            let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
            completion((200..<300).contains(code) ? .success(()) : .failure(.resolveFailed))
        }.resume()
    }

    private struct GuidResponse: Decodable { let guid: String }

    private static func fetchGuid(
        url: URL,
        method: String,
        body: Data?,
        completion: @escaping (Result<String, Web2AppError>) -> Void
    ) {
        var req = URLRequest(url: url)
        req.httpMethod = method
        if let body {
            req.httpBody = body
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        URLSession.shared.dataTask(with: req) { data, _, err in
            if let err { return completion(.failure(.network(err.localizedDescription))) }
            guard
                let data,
                let decoded = try? JSONDecoder().decode(GuidResponse.self, from: data),
                !decoded.guid.isEmpty
            else { return completion(.failure(.resolveFailed)) }
            completion(.success(decoded.guid))
        }.resume()
    }
}

File: android/src/main/kotlin/app/web2app/sdk/AttributionResolver.kt
package app.web2app.sdk

import org.json.JSONObject
import java.net.URLEncoder

/**
 * Резолв guid из carrier-token или email. Стабильные backend-контракты (POC-независимо):
 *  - token → `GET /public/handoff/resolve?code=<token>` → { guid } (WEB-433).
 *  - email → verified-resolve (WEB-431, In Review — контракт финализируется при мёрдже).
 */
internal class AttributionResolver(private val config: Web2AppConfig) {

    fun resolveToken(token: String, onResult: (Result<String>) -> Unit) {
        Http.io {
            val q = URLEncoder.encode(token, "UTF-8")
            onResult(parseGuid(Http.get("${config.baseUrl}/public/handoff/resolve?code=$q")))
        }
    }

    fun resolveEmail(email: String, onResult: (Result<String>) -> Unit) {
        Http.io {
            val body = JSONObject()
                .put("projectId", config.projectId)
                .put("email", email)
                .toString()
            // ⚠ WEB-431 (email-ядро, In Review): точный путь recovery request/verify финализируется.
            val resp = Http.postJson("${config.baseUrl}/public/handoff/email-recovery/verify", body)
            onResult(parseGuid(resp))
        }
    }

    private fun parseGuid(body: String?): Result<String> {
        val guid = body?.let { runCatching { JSONObject(it).optString("guid") }.getOrNull() }
        return if (!guid.isNullOrEmpty()) Result.success(guid)
        else Result.failure(IllegalStateException("resolve failed"))
    }
}
